### PR TITLE
test: Add python backend tests for the new histogram metric

### PR DESCRIPTION
**Pull request by yinggeh**
#### What does the PR do?
Tests histogram metric in custom_metrics.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [x] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [x] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [x] test

#### Related PRs:
https://github.com/triton-inference-server/vllm_backend/pull/56
https://github.com/triton-inference-server/python_backend/pull/374
https://github.com/triton-inference-server/core/pull/386

#### Where should the reviewer start?
n/a

#### Test plan:
n/a

- CI Pipeline ID:
17487728

#### Caveats:
n/a

#### Background
Customer requested histogram metrics from vLLM.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
n/a

### CODE REVIEW

File: qa/python_models/custom_metrics/model.py [hunk 1]
```diff
@@ -136,6 +226,27 @@ def test_gauge_e2e(self):
...
+        name = "test_histogram_e2e"
+        metric_family = pb_utils.MetricFamily(
+            name=name,
+            description="test metric histogram kind end to end",
+            kind=pb_utils.MetricFamily.HISTOGRAM,
+        )
+        labels = {"example1": "counter_label1", "example2": "counter_label2"}
+        buckets = [0.1, 1.0, 2.5, 5.0, 10.0]
+        metric = metric_family.Metric(labels=labels, buckets=buckets)
+        labels_str = 'example1="counter_label1",example2="counter_label2"'
```
Comment (rmccorm4) [inline qa/python_models/custom_metrics/model.py:239]: counter -> histogram

File: qa/python_models/custom_metrics/model.py [hunk 2]
```diff
@@ -136,6 +226,27 @@ def test_gauge_e2e(self):
         metrics = self._get_metrics()
         self.assertIn(pattern, metrics)
 
+    def test_histogram_e2e(self):
```
Comment (rmccorm4) [inline qa/python_models/custom_metrics/model.py:229]: Would be good to have some negative tests with no buckets provided to ensure no crash/hang:
```python
metric = metric_family.Metric(labels=labels)
metric = metric_family.Metric(labels=labels, buckets=[])
metric = metric_family.Metric(labels=labels, buckets=None)
```

File: qa/python_models/custom_metrics/model.py [hunk 3]
```diff
@@ -74,6 +74,96 @@ def _metric_api_helper(self, metric, kind):
...
+        )
+        self.assertIn(
+            histogram_str_builder(name, "bucket", labels, "0", le="5"), metrics
+        )
+        self.assertIn(
+            histogram_str_builder(name, "bucket", labels, "0", le="10"), metrics
+        )
+        self.assertIn(
+            histogram_str_builder(name, "bucket", labels, "0", le="+Inf"), metrics
+        )
```
Comment (rmccorm4) [inline qa/python_models/custom_metrics/model.py:117]:
```python
    for bucket in buckets:
        self.assertIn(
            histogram_str_builder(name, "bucket", labels, "0", le=str(bucket)), metrics
        )

File: qa/python_models/custom_metrics/model.py [hunk 4]
```diff
@@ -136,6 +226,27 @@ def test_gauge_e2e(self):
...
 
+    def test_histogram_e2e(self):
+        name = "test_histogram_e2e"
+        metric_family = pb_utils.MetricFamily(
+            name=name,
+            description="test metric histogram kind end to end",
+            kind=pb_utils.MetricFamily.HISTOGRAM,
+        )
+        labels = {"example1": "counter_label1", "example2": "counter_label2"}
+        buckets = [0.1, 1.0, 2.5, 5.0, 10.0]
```
Comment (rmccorm4) [inline qa/python_models/custom_metrics/model.py:237]: Testing duplicate values will be useful later on as well:
```python
duplicate_buckets = [1, 1, 1, 2, 2, 2]
metric = metric_family.Metric(labels=labels, buckets=duplicate_buckets)
# assert there is only one of each bucket value in metrics output, or capture
# error/exception, depending on how it behaves today
```

**Review by rmccorm4 — APPROVED** [review]
LGTM for cherry-pick purposes, but some comments for improving the tests later on.

**Review by rmccorm4 — DISMISSED** [review]

File: qa/python_models/custom_metrics/model.py (hunk 3 above)
Comment (yinggeh) [reply qa/python_models/custom_metrics/model.py:117]: Discussed offline.

**Comment by rmccorm4** [comment]
Please trigger a pipeline encapsulating all the latest changes so can feel confident in the CI impact when looking at the cherry-picks

File: qa/python_models/custom_metrics/model.py [hunk 5]
```diff
@@ -128,14 +218,56 @@ def test_gauge_e2e(self):
...
+        # Test non-ascending order buckets
+        with self.assertRaises(pb_utils.TritonModelException):
+            metric_family.Metric(labels={}, buckets=[2.5, 0.1, 1.0, 10.0, 5.0])
+
+        # Test duplicate value buckets
+        with self.assertRaises(pb_utils.TritonModelException):
+            metric_family.Metric(labels={}, buckets=[1, 1, 2, 5, 5])
+
+        # Test non-ascending order buckets
+        metric_family.Metric(labels={}, buckets=[])
```
Comment (rmccorm4) [inline qa/python_models/custom_metrics/model.py:269]: Should this be included? Didn't you find an issue with `buckets=[]`?

File: qa/python_models/custom_metrics/model.py (hunk 2 above)
Comment (yinggeh) [reply qa/python_models/custom_metrics/model.py:229]: Added all tests

File: qa/python_models/custom_metrics/model.py (hunk 4 above)
Comment (yinggeh) [reply qa/python_models/custom_metrics/model.py:238]: Updated

File: qa/python_models/custom_metrics/model.py (hunk 1 above)
Comment (yinggeh) [reply qa/python_models/custom_metrics/model.py]: Updated

File: qa/python_models/custom_metrics/model.py (hunk 5 above)
Comment (yinggeh) [reply qa/python_models/custom_metrics/model.py:275]: After the call, I notice something and investigated the error msg again. The mem_leak seems to be the false alarm and not the real error.
The real error was that `metric_family.Metric(labels={}, buckets=[])` successed and created the histogram with empty bucket. Then in the same test I ran the regular histogram e2e test and failed to create a new metric in the same `metric_family`. Not sure if it is expected behavior but I will follow it up.

```
AssertionError: 'test_histogram_e2e_bucket{example1="histogram_label1",example2="histogram_label2",le="0.1"} 0' not found in
# HELP test_histogram_e2e test metric histogram kind end to end
# TYPE test_histogram_e2e histogram
test_histogram_e2e_count{example1="histogram_label1",example2="histogram_label2"} 0
test_histogram_e2e_sum{example1="histogram_label1",example2="histogram_label2"} 0
test_histogram_e2e_bucket{example1="histogram_label1",example2="histogram_label2",le="+Inf"} 0'
```

**Comment by yinggeh** [comment]
Merged https://github.com/triton-inference-server/server/pull/7525 to the wrong branch. Need re-approval to merge to main.

File: qa/python_models/custom_metrics/model.py [hunk 6]
```diff
@@ -128,14 +218,62 @@ def test_gauge_e2e(self):
...
+
+        # Test non-ascending order buckets
+        with self.assertRaises(pb_utils.TritonModelException):
+            metric_family.Metric(labels={}, buckets=[2.5, 0.1, 1.0, 10.0, 5.0])
+
+        # Test duplicate value buckets
+        with self.assertRaises(pb_utils.TritonModelException):
+            metric_family.Metric(labels={}, buckets=[1, 1, 2, 5, 5])
+
+        # Test non-ascending order buckets
```
Comment (rmccorm4) [inline qa/python_models/custom_metrics/model.py:274]: Fix comment

Comment (yinggeh) [reply qa/python_models/custom_metrics/model.py]: Nice catch

**Review by rmccorm4 — APPROVED** [review]

**Pull request merged**